# Data Subscriptions: How to Consume a Data Subscription

---

{NOTE: }

* Batches of documents sent from a Subscription Task defined on the server are 
  consumed and processed by a subscription worker client.  
* The `subscription_worker` object, defined on the client, manages the communication 
  between the server and the client and processes the document batches sent from the server.  
* There are several ways to create and configure the SubscriptionWorker - see `SubscriptionWorkerOptions`.  

* In this page:
  * [`subscription_worker` lifecycle](../../../client-api/data-subscriptions/consumption/how-to-consume-data-subscription#subscription_worker-lifecycle)  
  * [Error handling](../../../client-api/data-subscriptions/consumption/how-to-consume-data-subscription#error-handling)  
  * [Worker interplay](../../../client-api/data-subscriptions/consumption/how-to-consume-data-subscription#worker-interplay)  
     * [Available Worker Strategies](../../../client-api/data-subscriptions/consumption/how-to-consume-data-subscription#available-worker-strategies)  
  * [Determining Which Workers a Subscription Will Serve](../../../client-api/data-subscriptions/consumption/how-to-consume-data-subscription#determining-which-workers-a-subscription-will-serve)  

{NOTE/}

---

{PANEL:`subscription_worker` lifecycle}
A `subscription_worker` object starts its life from being generated by the `store.subscriptions`:
{CODE:python subscription_open_simple@ClientApi\DataSubscriptions\DataSubscriptions.py /}

At this point, the worker has only got its configuration. No connection or processing happens at this moment. 
To start processing, the `run` method should be called. The Run method receives the batch processing logic that should be performed:
{CODE:python subscription_run_simple@ClientApi\DataSubscriptions\DataSubscriptions.py /}

From this point on, the subscription worker will start processing batches.  
If processing is aborted for any reason, the returned task (`subscription_runtime_task`) 
will be finished with an exception.  

{PANEL/}


{PANEL:Error handling}
There are two categories of errors that may occur during subscription processing:

{INFO:Internal mechanism errors}
Internal mechanism errors occur during the normal server-client communication between the worker and the server.  
If an unexpected error occurs, the worker will try to reconnect to the server.  
There are conditions in which the worker will cease its operation and will not try to reconnect:  

* The subscription does not exist or was deleted  
* Another worker took over the subscription (see connection strategy)
* The worker could not connect to any of the servers
* The worker could not receive the node in charge of the task (this can happen when there is no leader)
* Authorization exception
* Exception during connection establishment

{INFO/}

{INFO:User's batch processing logic unhandled exception}
Example:
{CODE:python throw_during_user_logic@ClientApi\DataSubscriptions\DataSubscriptions.py /}

If an exception is thrown, the worker will abort the current batch process. 
A worker can be configured to treat the thrown exception in either of the following two ways:  

* By default, the worker will wrap the thrown exception with a `SubscriberErrorException` exception 
  and rethrow it, terminating the subscription execution without acknowledging progress or retrying.  
  The task returned by the `run` function will be terminated with an erroneous state, throwing 
  a `SubscriberErrorException` exception.  

* If `SubscriptionWorkerOptions`'s value `IgnoreSubscriberErrors` is set to True, the erroneous 
  batch will get acknowledged without retrying and the next batches will continue processing. 
{INFO/}

{INFO: Reconnecting}
The above cases describe situations in which a worker tries to reconnect with the server.  
There are two key `SubscriptionWorkerOptions` fields controlling this state:

* `TimeToWaitBeforeConnectionRetry` - Worker 'sleep' period before trying to reconnect.  
* `MaxErroneousPeriod` - Maximum time that the worker is allowed to be in an erroneous state.  
  When this time elapses, the worker will stop trying to reconnect.  
{INFO/}

{INFO: OnUnexpectedSubscriptionError}
`OnUnexpectedSubscriptionError` is the event raised when a connection failure occurs 
between the subscription worker and the server and it throws an unexpected exception. 
When this occurs, the worker will automatically try to reconnect again. This event is 
useful for logging these unexpected exceptions.
{INFO/}

{PANEL/}

{PANEL: Worker interplay}

* Subscription workers are configured with a **strategy**, that determines whether 
  they can connect their subscription **concurrently**, or only one at a time.  
* If a one-at-a-time strategy is chosen, it also determines how the workers 
  interact with each other to resolve which will connect the subscription.  

{NOTE: }

The strategy is configured by the `SubscriptionWorkerOptions` `strategy` field.  
The strategy field is the `SubscriptionOpeningStrategy` enum.  

| `SubscriptionOpeningStrategy` | |
| ------------- | ------------- |
| `OPEN_IF_FREE` | Connect if no other worker is connected |
| `WAIT_FOR_FREE` | Wait for currently connected worker to disconnect |
| `TAKE_OVER` | Take over the connection |
| `CONCURRENT` | Connect concurrently |

{NOTE/}

## Available Worker Strategies

---

### One Worker Per Subscription Strategies 

The following three strategies allow only a **single worker to connect 
the subscription at any given time**, and determine what happens when one 
worker is connected and another tries to connect.  

* `SubscriptionOpeningStrategy.OPEN_IF_FREE`  
  The server will allow the worker to connect only if there is currently 
  no other connected worker.  
  If there is an existing connection, the incoming worker will throw a 
  `SubscriptionInUseException`.  
* `SubscriptionOpeningStrategy.WAIT_FOR_FREE`  
  If the worker currently cannot open the subscription because it is used 
  by another worker, it will wait for the previous worker to disconnect and 
  connect only then.  
  This is useful in worker failover scenarios, where there is one active 
  worker and another already waits to take its place.  
* `SubscriptionOpeningStrategy.TAKE_OVER`  
  The server will allow an incoming connection to overthrow an existing one, 
  according to the existing connection strategy.  
   * If the existing connection **does not** have a `TAKE_OVER` strategy:  
     The incoming connection will take over, causing the existing 
     connection to throw a `SubscriptionInUseException`.  
   * If the existing connection **has** a `TAKE_OVER` strategy:  
     The incoming connection will throw a `SubscriptionInUseException` exception.  

---

### Concurrent Strategy 

* `SubscriptionOpeningStrategy.CONCURRENT`  
  Multiple workers of the same subscription are allowed to connect it simultaneously.  
  Read more about concurrent subscriptions [here](../../../client-api/data-subscriptions/concurrent-subscriptions).  

{PANEL/}

{PANEL: Determining Which Workers a Subscription Will Serve}

{NOTE: }
The **strategy used by the first worker connecting to a subscription** will determine 
which additional workers this subscription will be able to serve until all worker connections
are dropped.  
{NOTE/}

* A subscription that serves one or more [CONCURRENT](../../../client-api/data-subscriptions/consumption/how-to-consume-data-subscription#concurrent-strategy) 
  workers, **can serve only other concurrent workers** until all worker connections are dropped.  
  If a worker that uses a [One Worker Per Subscription](../../../client-api/data-subscriptions/consumption/how-to-consume-data-subscription#one-worker-per-subscription-strategies) 
  strategy attempts to connect it -  
   * The connection attempt will be rejected.  
   * `SubscriptionInUseException` will be thrown.  

* A subscription that serves a worker that uses a
  [One Worker Per Subscription](../../../client-api/data-subscriptions/consumption/how-to-consume-data-subscription#one-worker-per-subscription-strategies) 
  strategy, **cannot** serve 
  [CONCURRENT](../../../client-api/data-subscriptions/consumption/how-to-consume-data-subscription#concurrent-strategy) 
  workers until the worker's connection is dropped.  
  If a concurrent worker attempts to connect it -  
   * The connection attempt will be rejected.  
   * `SubscriptionInUseException` will be thrown.  

{PANEL/}

## Related Articles  

### Data Subscriptions  

- [What are Data Subscriptions](../../../client-api/data-subscriptions/what-are-data-subscriptions)  
- [How to Create a Data Subscription](../../../client-api/data-subscriptions/creation/how-to-create-data-subscription)  
